src/lib/provable/test/dynamic-array.unit-test.ts

```diff
@@ -0,0 +1,312 @@
+import { UInt8, Field, ZkProgram } from 'o1js';
+import { DynamicArray } from '../dynamic-array.js';
+import { assert } from '../gadgets/common.js';
+
+function expectThrows(fn: () => void | Promise<void>, msg: string) {
+  let threw = false;
+  try {
+    fn();
+  } catch {
+    threw = true;
+  }
+  assert(threw, msg);
+}
+
+let List = ZkProgram({
+  name: 'dynamicarrays',
+  methods: {
+    pushAndPop: {
+      privateInputs: [],
+      async method() {
+        // Define classes of dynamic arrays for specific provable types
+        class Bytestring extends DynamicArray(UInt8, { capacity: 8 }) {}
+
+        // Initialize an empty dynamic array
+        let bytes = new Bytestring();
+
+        // Initialization should produce right capacity and length
+        assert(bytes.capacity === 8);
+        assert(bytes.length.equals(new Field(0)));
+
+        // Pushing elements should increase length but keep capacity
+        bytes.push(new UInt8(1));
+        assert(bytes.length.equals(new Field(1)));
+        assert(bytes.capacity === 8);
+        assert(bytes.get(new Field(0)).value.equals(new Field(1)));
+
+        // Popping elements should decrease length but keep capacity
+        bytes.pop();
+        assert(bytes.length.equals(new Field(0)));
+        assert(bytes.capacity === 8);
+
+        // Cannot push more elements than the capacity
+        for (let i = 0; i < 8; i++) {
+          bytes.push(new UInt8(i));
+        }
+        expectThrows(() => {
```

Comment:
This is tricky to test in a circuit because we can't branch the logic of a `try .. catch` block